index.js

```diff
@@ -1,85 +1,108 @@
-const path = require('path');
-const urlMod = require('url');
-const base64Stream = require('base64-stream');
-const parseCookiePhantomjs = require('parse-cookie-phantomjs');
-const phantomBridge = require('phantom-bridge');
-const byline = require('byline');
-
-const handleCookies = (cookies, url) => {
-	const parsedUrl = urlMod.parse(url);
-
-	return (cookies || []).map(x => {
-		const ret = typeof x === 'string' ? parseCookiePhantomjs(x) : x;
-
-		if (!ret.domain) {
-			ret.domain = parsedUrl.hostname;
-		}
-
-		if (!ret.path) {
-			ret.path = parsedUrl.path;
-		}
-
-		return ret;
-	});
+const fileUrl = require('file-url');
+const isUrl = require('is-url-superb');
+const puppeteer = require('puppeteer');
+const toughCookie = require('tough-cookie');
+
+const parseCookie = cookie => {
+	if (typeof cookie === 'object') {
+		return cookie;
+	}
+
+	const ret = toughCookie.parse(cookie).toJSON();
+	ret.name = ret.key;
+	return ret;
 };
 
-module.exports = (url, size, opts) => {
+const hideElement = element => {
+	element.style.visibility = 'hidden';
+};
+
+const getBoundingClientRect = element => {
+	const {height, width, x, y} = element.getBoundingClientRect();
+	return {height, width, x, y};
+};
+
+module.exports = async (url, opts) => {
 	opts = Object.assign({
-		delay: 0,
-		scale: 1,
-		format: 'png'
+		cookies: [],
+		fullPage: true,
+		hide: [],
+		width: 1920,
+		height: 1080
 	}, opts);
 
-	const args = Object.assign(opts, {
-		url,
-		width: size.split(/x/i)[0] * opts.scale,
-		height: size.split(/x/i)[1] * opts.scale,
-		cookies: handleCookies(opts.cookies, url),
-		format: opts.format === 'jpg' ? 'jpeg' : opts.format,
-		css: /\.css$/.test(opts.css) ? fs.readFileSync(opts.css, 'utf8') : opts.css,
-		script: /\.js$/.test(opts.script) ? fs.readFileSync(opts.script, 'utf8') : opts.script
-	});
-
-	const cp = phantomBridge(path.join(__dirname, 'stream.js'), [
-		'--ignore-ssl-errors=true',
-		'--local-to-remote-url-access=true',
-		'--ssl-protocol=any',
-		JSON.stringify(args)
-	]);
-
-	const stream = base64Stream.decode();
-
-	process.stderr.setMaxListeners(0);
-
-	cp.stderr.setEncoding('utf8');
-	cp.stdout.pipe(stream);
-
-	byline(cp.stderr).on('data', data => {
-		data = data.trim();
-
-		if (/ phantomjs\[/.test(data)) {
-			return;
-		}
-
-		if (/http:\/\/requirejs.org\/docs\/errors.html#mismatch/.test(data)) {
-			return;
-		}
-
-		if (data.startsWith('WARN: ')) {
-			stream.emit('warning', data.replace(/^WARN: /, ''));
-			stream.emit('warn', data.replace(/^WARN: /, '')); // TODO: deprecate this event in v5
-			return;
-		}
-
-		if (data.length > 0) {
-			const err = new Error(data);
-			err.noStack = true;
-			cp.stdout.unpipe(stream);
-			stream.emit('error', err);
-		}
-	});
-
-	return stream;
+	const uri = isUrl(url) ? url : fileUrl(url);
+	const {
+		cookies, crop, format, headers, height, hide, keepAlive, password, scale,
+		script, selector, timeout, transparent, userAgent, username, width
+	} = opts;
```

Comment:
This feels like overusing destructuring just for the sake of it. And using `opts.cookies` in the code makes it more explicit it's coming from a user option.